ops/model.py

```diff
@@ -663,3 +663,10 @@ def network_get(self, endpoint_name, relation_id=None):
             if 'relation not found' in str(e):
                 raise RelationNotFoundError() from e
             raise
+
+    def add_metric(self, metrics, labels=None):
```

Comment:
I think we need some validation that keys/values are what we expect. If they contain commas, equal signs, spaces (maybe?), it'll create unexpected commands. What's the actual format accepted by juju for such data?

Reply:
Juju splits the passed labels string using a comma as a separator:
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/jujuc/add-metric.go#L68-L70

For metrics themselves each arg is split via `=` and whitespace of a key (to the left of `=`) is trimmed.
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/jujuc/add-metric.go#L71-L75
https://github.com/juju/utils/blob/c746c6e86f4fb2a04bc08d66b7a0f7e900d9cbab/keyvalues/keyvalues.go#L22-L41

I don't see any particular validation for metric keys which might be a Juju-side issue:

https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/jujuc/add-metric.go#L76-L82
https://github.com/juju/charm/blob/v6/metrics.go#L78-L120
https://github.com/juju/charm/blob/v6/metrics_test.go#L105-L173

Besides the builtin `juju` [prefix usage](https://github.com/juju/charm/blob/v6/metrics.go#L91-L95), the only restriction I can see is that metrics need to be valid YAML keys per the spec
https://yaml.org/spec/1.2/spec.html#Characters

Reply:
The point was probably not clear. Imagine you have a comma in a key, or an equal sign.. what happens?

Given that, we need to restrict the vocabulary. What is the exact vocabulary accepted by juju, so we're not diverging? I assume it's not a completely lenient string handling.

Reply:
There is no regex in Juju to validate metric keys as far as I can see. They will be valid YAML keys and the keys passed via add-metrics will be checked against metrics.yaml. This means that having a comma or an equal sign for a key would work.
https://github.com/juju/charm/blob/v6/metrics.go#L111-L120

The current equal sign parsing logic doesn't seem to be right either as it will accept multiple equal signs in a key:
https://github.com/juju/utils/blob/c746c6e86f4fb2a04bc08d66b7a0f7e900d9cbab/keyvalues/keyvalues.go#L27-L30
https://play.golang.org/p/k6LBao7ByGH

Reply:
`SplitN(.., 2)` means it takes the first equal sign and splits it. So

`key=value=with=equals`

would give you `('key', 'value=with=equals')`

So while you could technically define a metric with an '=' in its name, it makes parsing ambiguous and would be a bad idea.

The test suite does a bunch of checks about whether you have whitespace around the '='
https://github.com/juju/utils/blob/c746c6e86f4fb2a04bc08d66b7a0f7e900d9cbab/keyvalues/keyvalues_test.go

but it doesn't explicit test things like `key space = value space` => `('key space', 'value space')` though I'm pretty sure that is how it would parse them.

We can chose how strict we want to be to assist users. Giving them an error if the key contains an '=' or ',' and the value if it contains a ',' would be reasonable.
For actual metrics, though, the value has to be a valid float:
https://github.com/juju/juju/blob/d05cd0783d4027e7a77cb08ffc575aaf71bd9f32/worker/metrics/spool/metrics.go#L220
```
func (m *JSONMetricRecorder) validateMetric(key, value string) error {
	if !m.IsDeclaredMetric(key) {
		return errors.Errorf("metric key %q not declared by the charm", key)
	}
	// The largest number of digits that can be returned by strconv.FormatFloat is 24, so
	// choose an arbitrary limit somewhat higher than that.
	if len(value) > 30 {
		return fmt.Errorf("metric value is too large")
	}
	fValue, err := strconv.ParseFloat(value, 64)
	if err != nil {
		return fmt.Errorf("invalid value type: expected float, got %q", value)
	}
	if fValue < 0 {
		return fmt.Errorf("invalid value: value must be greater or equal to zero, got %v", value)
	}
	return nil
}
```

I'm actually surprised it has to be a positive metric, I probably wouldn't hard code that at the charm level. (It might make sense for metrics-that-you-use-for-billing to never be negative, but general metrics don't.)

Reply:
Addressed this in https://github.com/canonical/operator/pull/140/commits/d62f516e189b8af8771b607632d96e0c6228e5c6 per our discussion.

We will also follow-up with @cmars about the expected format for metrics and labels.

Reply:
After talking to @cmars we got some statistics about the metrics in-use today.

There are metrics that have dashes and underscores.

Examples:
- patches-applied
- custom-channels
- nodes-gpu
- current_uptime
- current_pools
- storage_pool_total

Additionally, Juju docs document a `juju-units` metric.
https://discourse.jujucharms.com/t/metric-collecting-charms/1125